docker/Dockerfile

```diff
@@ -45,33 +58,37 @@ RUN \
     zlib1g-dev \
     && rm -rf /var/lib/apt/lists/*
 
-# Copy just what we need to pip install
-COPY MANIFEST.in README.rst setup.py /synapse/
-COPY synapse/__init__.py /synapse/synapse/__init__.py
-COPY synapse/python_dependencies.py /synapse/synapse/python_dependencies.py
+WORKDIR /synapse
+
+# Copy just what we need to run `poetry install`
+COPY pyproject.toml poetry.lock README.rst /synapse/
+
+# Install to the Python installation which hosts `pip`. In this case, it's the system
+# Python.
+ENV POETRY_VIRTUALENVS_IN_PROJECT=true \
+    POETRY_VIRTUALENVS_CREATE=true \
+    POETRY_HOME=/opt/poetry
 
 # To speed up rebuilds, install all of the dependencies before we copy over
-# the whole synapse project so that we this layer in the Docker cache can be
+# the whole synapse project, so that this layer in the Docker cache can be
 # used while you develop on the source
-#
-# This is aiming at installing the `install_requires` and `extras_require` from `setup.py`
-RUN --mount=type=cache,target=/root/.cache/pip \
-  pip install --prefix="/install" --no-warn-script-location \
-    /synapse[all]
+RUN --mount=type=cache,target=/opt/poetry/artifacts \
+    --mount=type=cache,target=/opt/poetry/.cache/pypoetry/cache \
+  /root/.local/bin/poetry install --no-dev --no-root --no-interaction --no-ansi --extras all
 
-# Copy over the rest of the project
+# Copy over the synapse source code.
```

Comment:
@richvdh [suggested](https://github.com/matrix-org/synapse/pull/12337/files#r842766766):

```suggestion
# Copy over the rest of the synapse source code.
```

It's nitpicking, but the only thing we copied over previously was `pyproject.toml`, `poetry.lock` and `README.rst`: to my mind, we haven't copied any source code yet!

Reply:
well, ok, But if you're using that definition, then when we copy the whole of `synapse`, we're not just copying source code!

As you say though, nitpicking. I don't have strong feelings.

Reply:
Nor I. Went with your wording in 92c9a60bd,